pkg/assembler/backends/inmem/osv.go

```diff
@@ -135,7 +135,8 @@ func (c *demoClient) Osv(ctx context.Context, filter *model.OSVSpec) ([]*model.O
 		}
 		osv, err := c.buildOsvResponse(uint32(id), filter)
 		if err != nil {
-			return nil, err
+			// not found
+			return nil, nil
```

Comment:
This is weird, we are losing the error here.

Reply:
The error here is if the id is not found. The correct thing to do is return an empty response, not error.

Reply:
We are returning both `nil, err` and `nil, nil` from the function: https://github.com/guacsec/guac/blob/fc490e080b27234538d8bf2afa6b7471fc8322ef/pkg/assembler/backends/inmem/osv.go#L199-L205

I think we actually need to process that error and return nil, nil from there on not found and propagate any other error upstream

Reply:
ByID() returns error if the id is not in the map or is not the type we are looking for. The parseuint error is checked further up (and we now have a unit test for that).

Reply:
I think we should still report an error if the id does not map to the expected type. Can we make `byID` return that error and `nil` if not found instead of erroring there?

Reply:
> I think we should still report an error if the id does not map to the expected type

I'd expect 

```
{
  packages(pkgSpec: { ID: 10}) {
    ...
  }
}

```
To return an empty list if there is no package with ID 10. That would be the same behavior if you specified a name or other field that didn't match any package.

Reply:
Yes, but we still need to signal an error if the database is corrupted and we get an unexpected type out of ByID when passing in an ID from a backedge and getting something else.

We can make `ByID` not return an error and ignore the corruption scenario above, or we can add a flag to it to signal when the type mismatch should be treated as not found instead of a data corruption.

Reply:
Right, most calls to ByID expect to find something (backedge), and will throw on error. In this case we know that an error from ByID means not found, and because the ID was part of the query we don't throw, but return an empty result.

Reply:
Yes, but I saying we should mark it explicit in the API that we want to `ByID` to not error here / that we want this specific error from `buildOSVResponse` to be ignored. Otherwise, future changes in `buildOSVResponse` might raise a new error which we are silently ignoring here when we shouldn't.

Reply:
>  Otherwise, future changes in buildOSVResponse might raise a new error which we are silently ignoring here when we shouldn't.

Yes, I agree this is best practice. We have quite a few of this same pattern, I'll take a look